### PR TITLE
fixes/filter-bump

### PR DESCRIPTION
**Pull request by decanus**

### CODE REVIEW

**Comment by decanus** [comment]
@oskarth this is part of the first fix @dryajov and I were working on. It seems as though the 3 issues may relate to libp2p and @dryajov needs to investigate further in that codebase. The changes in here will be required though as they fix parts of the filter protocol

**Comment by dryajov** [comment]
So, looking at it some more and the issues is still the way that the handler is implemented - https://github.com/status-im/nim-waku/blob/bump-submodules/waku/protocol/v2/waku_filter.nim#L102-L109. One way to look at the handler method is to think of it as a `main` function/method (if familiar with languages like C/C++ or Java, etc..), ie an entry point. Once the the handler exits, execution is finished and all resources are freed. 

The usual thing to do is to have a read loop that constantly reads from the stream and does something within the message, because storing the stream for later use will not work because it will be destroyed as soon as the handler exists. 

Looking at the filter test and the handler implementation, it would seem like you'd probably want to notify the subscribers as soon as the proper message arrives, so maybe something like this:

(note the `while not conn.atEof`)
```nim
 proc handle(conn: Connection, proto: string) {.async, gcsafe, closure.} =
   while not conn.atEof:
      var message = await conn.readLp(64*1024)
      var res = FilterRequest.init(message)
      # DO SOMETHING WITH THE MESSAGE
```

If you need to store the messages per peer and then resend them at a later time, the next best thing to do is to store the `PeerInfo` of the peer and then when the time comes get a new connection or stream to that peer and send the messages:

```nim
 proc handle(conn: Connection, proto: string) {.async, gcsafe, closure.} =
    var message = await conn.readLp(64*1024)
    var res = FilterRequest.init(message)
    if res.isErr:
      return

    ws.subscribers.add(Subscriber(peer: conn.peerInfo, filter: res.value))
    # @TODO THIS IS A VERY ROUGH EXPERIMENT
```

Then the notify handler becomes something like:

```nim
  proc handle(msg: Message) {.async.} =
    var futures = newSeq[Future[void]]()

    for subscriber in proto.subscribers:
        if subscriber.filter.topic in msg.topicIDs:
          # @TODO PROBABLY WANT TO BATCH MESSAGES
          let conn = switch.dial(subscriber.peer, codec)
          futures.add(conn.writeLp(MessagePush(messages: @[msg]).encode().buffer))

    await allFutures(futures)
```

The second and third examples imply passing and storing the switch to the protocol to then use it to acquire a connection and/or stream when required. 

The first example would do it all reusing the same incoming stream.

**Comment by oskarth** [comment]
Why are we looking at filter now ? We explicitly said several times the priority is not that protocol, but the node and relay. This was mentioned in the original PR twice https://github.com/status-im/nim-waku/pull/157 in #waku Discord and in PM to you @decanus. I would've expected the single failing test in https://github.com/status-im/nim-waku/pull/157 to be the focus on any investigation.

**Review by oskarth — APPROVED** [review]
As for the actual changes they generally seem fine to me, but it is a draft so not sure we can build on it. Would be better to just push directly to the existing branch, and better still to focus on the critical path.

File: Makefile [hunk 1]
```diff
@@ -127,7 +127,7 @@ docker-push:
 
 # builds and runs the test suite
 test: | build deps
-	$(ENV_SCRIPT) nim test $(NIM_PARAMS) waku.nims
+	# $(ENV_SCRIPT) nim test $(NIM_PARAMS) waku.nims
```
Comment (oskarth) [inline Makefile:130]: This is on purpose, if you remove this I'm not sure waku v1 tests are run. Please unstage

**Review by oskarth — CHANGES_REQUESTED** [review]

**Comment by oskarth** [comment]
> This is on purpose, if you remove this I'm not sure waku v1 tests are run. Please unstage

I'll do this in the branch instead, easier.

**Comment by oskarth** [comment]
This PR was merged prematurely, please see https://github.com/status-im/nim-waku/pull/157#issuecomment-693142459

Either 1) change target to master or 2) push to that branch without conflicts

**Comment by kdeme** [comment]
> Why are we looking at filter now ? We explicitly said several times the priority is not that protocol, but the node and relay. 

Also don't fully understand this here. I think it is pretty clear that WakuFilter is incomplete, as in the whole message dispatching part is missing. But I don't think this was the core issue at hand, was it?

**Pull request merged**